test/cpu_fusion.cpp

```diff
@@ -1213,13 +1215,126 @@ TEST(cpu_fusion, backwards_maxpool_with_indices_n4_c1_hw4_2x2_max)
         pass_manager.run_passes(df);
     }
 
     backend->call(df, {output}, {input, ep});
     ASSERT_TRUE(read_vector<float>(output) == expected);
 }
+
+static std::shared_ptr<ngraph::Function> make_forward_function()
+{
+    Shape shape_a{10, 3, 28, 28};
+    auto input = std::make_shared<op::Parameter>(element::f32, shape_a);
+    Shape window_shape{2, 2};
+    auto max_pool = std::make_shared<op::MaxPool>(input, window_shape);
+    auto neg = std::make_shared<op::Negative>(max_pool);
+    auto absn = std::make_shared<op::Abs>(max_pool);
+    return std::make_shared<Function>(NodeVector{max_pool, neg, absn}, op::ParameterVector{input});
+}
+
+static std::pair<std::shared_ptr<ngraph::Function>, std::vector<std::shared_ptr<ngraph::Node>>>
+    make_backward_function(std::shared_ptr<ngraph::Function> f)
+{
+    // get parameters
+    std::vector<std::shared_ptr<ngraph::op::Parameter>> back_parameters = f->get_parameters();
+
+    ngraph::NodeVector adjoints;
+    ngraph::NodeVector outputs;
+    for (auto Y : f->get_results())
+    {
+        // Get the output
+        // Create the Adjoint
+        auto C = std::make_shared<ngraph::op::Parameter>(Y->get_element_type(), Y->get_shape());
+        outputs.push_back(Y);
+        adjoints.push_back(C);
+    }
+
+    ngraph::autodiff::Adjoints adjoint{outputs, adjoints};
+
+    // Perform autodiff
+    std::vector<std::shared_ptr<Node>> dYdXs(back_parameters.size());
+    transform(back_parameters.begin(),
+              back_parameters.end(),
+              dYdXs.begin(),
+              [&adjoint](const std::shared_ptr<Node>& X) { return adjoint.backprop_node(X); });
+
+    // create the backward function
+    std::vector<std::shared_ptr<ngraph::op::Parameter>> param_adjoints;
+    for (auto n : adjoints)
+        param_adjoints.push_back(std::dynamic_pointer_cast<ngraph::op::Parameter>(n));
+    back_parameters.insert(back_parameters.begin(), param_adjoints.begin(), param_adjoints.end());
+
+    return {std::make_shared<ngraph::Function>(dYdXs, back_parameters), adjoints};
+}
+
+void optimize_graph(std::shared_ptr<ngraph::Function>& f, std::shared_ptr<ngraph::Function> bf)
+{
+    // start by removing excess reshapes
+    NodeVector nv_cwi;
+    ngraph::pass::Manager pass_manager;
+    pass_manager.register_pass<ngraph::pass::ReshapeElimination>();
+    pass_manager.register_pass<ngraph::pass::ReshapeElimination>();
+    pass_manager.register_pass<runtime::cpu::pass::CPUWorkspaceInsertion>(nv_cwi);
+    pass_manager.register_pass<pass::VisualizeTree>("before.fprop_cache.pdf");
+
+    pass_manager.run_passes(f);
+    pass_manager.run_passes(bf);
+    if (nv_cwi.size() > 0)
+    {
+        NodeVector new_outputs;
+        for (auto r : f->get_results())
+        {
+            new_outputs.push_back(r->get_argument(0));
+        }
+
+        new_outputs.insert(new_outputs.end(), nv_cwi.begin(), nv_cwi.end());
+        f = std::make_shared<ngraph::Function>(new_outputs, f->get_parameters());
```

Comment:
I think this should be a new Function, instead of overwriting f. Doesn't matter when reimplementing

Reply:
@mbrookhart indeed, this is pretty much copy-pasted from `ngraph-mxnet`. However, this is ngraph's test code and making `ngraph-mxnet` rely/call the *test* code doesn't quite give me warm fuzzy feelings. It precludes us making arbitrary changes to this test, extending it because of an introduced dependency on `ngraph-mxnet`. I'd rather have the two separate.

Reply:
@mbrookhart I wouldn't care much until I have to test something in the context of MxNet bridge at which point I'll pull yr changes manually. You don't seem to be updating this logic too often and my need to use it is even less frequent. Also, usually you introduce very minor changes to this particular logic which would be very really to port.

Reply:
:) Okay

Reply:
:yayfox